### PR TITLE
feat(readonly): Adding test cases for uzfs pool readonly support

### LINKED ISSUE #2937: cstor-pool process crashes in loop due to pool being full
# Description
If cstor pool space gets exhausted then there is no possible way to extend/reclaim the storage space.

## Expected Behavior
If cstor pool space gets consumed up-to certain percentage then it should not allow more ios until pool is expanded.

### PR DESCRIPTION
**Pull request by mynktl**
changes:
- Adding support to make uzfs pool readonly
To set pool readonly:
```
zpool  set io.openebs:readonly=on testp
```

To unset pool readonly:
```
zpool  set io.openebs:readonly=off testp
```

Test cases added for following scenarios:
- If pool is set readonly then zrepl shouldn't scan this zvol, created on the pool, to make connection with target
- if zvol was already connected while setting readonly mode then write/snapshot{prepare/create} will fail, but read will be allowed.

Fixes: https://github.com/openebs/openebs/issues/2937
Signed-off-by: mayank <mayank.patel@mayadata.io>

### CODE REVIEW

File: include/sys/spa_impl.h [hunk 1]
```diff
@@ -297,6 +297,9 @@ struct spa {
 	taskq_t		*spa_prefetch_taskq;	/* Taskq for prefetch threads */
 	uint64_t	spa_multihost;		/* multihost aware (mmp) */
 	mmp_thread_t	spa_mmp;		/* multihost mmp thread */
+#ifdef	_UZFS
```
Comment (vishnuitta) [inline include/sys/spa_impl.h:300]: we hadn't touched spa till now.. may be we never got that need.. can we also explore different way? if not possible, this is fine

File: include/sys/fs/zfs.h [hunk 2]
```diff
@@ -229,6 +229,7 @@ typedef enum {
 	ZPOOL_PROP_TNAME,
 	ZPOOL_PROP_MAXDNODESIZE,
 	ZPOOL_PROP_MULTIHOST,
+	ZPOOL_PROP_UZFS_READONLY,
```
Comment (vishnuitta) [inline include/sys/fs/zfs.h:233]: this as well should go under _UZFS macro

File: module/zfs/spa.c [hunk 3]
```diff
@@ -729,6 +743,21 @@ spa_prop_set(spa_t *spa, nvlist_t *nvp)
...
+			if (strcmp(val, "on") == 0) {
+				spa->readonly = B_TRUE;
+			} else if (strcmp(val, "off") == 0) {
+				spa->readonly = B_FALSE;
+			} else {
+				return (EINVAL);
+			}
+			dmu_objset_find(spa->spa_name, uzfs_zpool_rdonly_cb,
+			    val, DS_FIND_CHILDREN);
+		}
```
Comment (vishnuitta) [inline module/zfs/spa.c:772]: we need to do continue here

Comment (vishnuitta) [reply module/zfs/spa.c:772]: do we need to have need_sync as well to B_TRUE?

File: module/zfs/spa.c [hunk 4]
```diff
@@ -2141,6 +2170,28 @@ spa_load_verify(spa_t *spa)
...
 
+#ifdef	_UZFS
+static void
+uzfs_get_readonly_prop(spa_t *spa)
+{
+	char val[4];	// max strlen("off")
+	int err;
+
+	err = zap_lookup(spa->spa_meta_objset, spa->spa_pool_props_object,
+	    zpool_prop_to_name(ZPOOL_PROP_UZFS_READONLY), 1, 4, &val);
```
Comment (pawanpraka1) [inline module/zfs/spa.c]: use sizeof(val)

File: module/zfs/spa.c [hunk 5]
```diff
@@ -635,7 +636,17 @@ spa_prop_validate(spa_t *spa, nvlist_t *props)
...
+#ifdef	_UZFS
+		case ZPOOL_PROP_UZFS_READONLY:
+			if ((error = nvpair_value_string(elem, &strval)) != 0)
+				break;
 
+			if (strcmp(strval, "on") != 0 &&
+			    strcmp(strval, "off") != 0) {
+				error = SET_ERROR(EINVAL);
+				break;
+			}
```
Comment (pawanpraka1) [inline module/zfs/spa.c:647]: better to put break for this case.

File: module/zfs/spa.c [hunk 6]
```diff
@@ -729,6 +743,23 @@ spa_prop_set(spa_t *spa, nvlist_t *nvp)
...
+		if (prop == ZPOOL_PROP_UZFS_READONLY) {
+			VERIFY(nvpair_value_string(elem, &val) == 0);
+			if (strcmp(val, "on") == 0) {
+				spa->readonly = B_TRUE;
+			} else if (strcmp(val, "off") == 0) {
+				spa->readonly = B_FALSE;
+			} else {
+				return (EINVAL);
+			}
+			dmu_objset_find(spa->spa_name, uzfs_zpool_rdonly_cb,
```
Comment (pawanpraka1) [inline module/zfs/spa.c]: we don't need to set this on very child if it is set on the pool.

File: module/zfs/spa.c [hunk 7]
```diff
@@ -729,6 +743,23 @@ spa_prop_set(spa_t *spa, nvlist_t *nvp)
...
+			    val, DS_FIND_CHILDREN);
+			need_sync = B_TRUE;
```
Comment (pawanpraka1) [inline module/zfs/spa.c]: lets avoid need_sync if it is already set.

Comment (mynktl) [reply module/zfs/spa.c]: I think that is not needed, right?

File: module/zfs/spa.c (hunk 4 above)
Comment (mynktl) [reply module/zfs/spa.c]: ok

File: module/zfs/spa.c (hunk 5 above)
Comment (mynktl) [reply module/zfs/spa.c:647]: ok

**Review by pawanpraka1 — DISMISSED** [review]
looks good.

**Review by vishnuitta — DISMISSED** [review]
changes are good

**Review by vishnuitta — APPROVED** [review]
changes are good

**Pull request merged**